src/analysis/inner_product_space/finite_dimensional.lean

```diff
@@ -0,0 +1,229 @@
+-/
+
+variables {V : Type*} [inner_product_space ℂ V]
+local notation `P`U := (orthogonal_projection U)
+local notation `P`U`ᗮ` := (orthogonal_projection Uᗮ)
+
+/-- `U` is `T` invariant: `∀ u : V`, if `u ∈ U` then `T u ∈ U`-/
+def invariant_subspace (U : submodule ℂ V) (T : V →L[ℂ] V) :
+ Prop := U ≤ U.comap T
+lemma invariant_subspace_def (U : submodule ℂ V) (T : V →L[ℂ] V) :
+ invariant_subspace U T ↔ (T '' U ⊆ U)
+  := by simp only [set.image_subset_iff]; refl
+
+/-- `U` is `T` invariant if and only if `(P U) * T * (P U) = T * (P U)` -/
+lemma subspace_is_invariant_iff_ortho_proj_mul_T_mul_ortho_proj_eq_T_mul_ortho_proj
+  [finite_dimensional ℂ V] (U : submodule ℂ V) (T : V →L[ℂ] V) :
+  (invariant_subspace U T) ↔ ∀ x : V, ↑((P U) (T ↑((P U) x))) = (T ↑((P U) x)) :=
+begin
+  split,
+  { intros h x,
+    obtain ⟨w, hw, v, hv, hvw⟩ := submodule.exists_sum_mem_mem_orthogonal U x,
+    rw [hvw, map_add,
+      orthogonal_projection_mem_subspace_orthogonal_complement_eq_zero hv,
+      add_zero, orthogonal_projection_eq_self_iff.mpr hw],
+    exact orthogonal_projection_eq_self_iff.mpr (h hw) },
+  { intros h u h_1,
+    rw [submodule.mem_comap,
+      ← orthogonal_projection_eq_self_iff, ← orthogonal_projection_eq_self_iff.mpr h_1, h] }
+end
+
+/-- `U,Uᗮ` are `T` invariant if and only if `(P U) * T = T * (P U)` -/
+lemma U_and_U_bot_are_T_invariant_iff_ortho_proj_mul_T_eq_T_mul_ortho_proj
+ [finite_dimensional ℂ V] (U : submodule ℂ V) (T : V →L[ℂ] V) :
+ (invariant_subspace U T ∧ invariant_subspace Uᗮ T) ↔ ∀ x : V, ↑((P U) (T x)) = (T ↑((P U) x)) :=
+   begin
+     simp only [subspace_is_invariant_iff_ortho_proj_mul_T_mul_ortho_proj_eq_T_mul_ortho_proj],
+     have : ∀ x : V,
+      ↑((P Uᗮ) x) = ((continuous_linear_map.id ℂ V) x) - ↑((P U) x)
+        := λ x, by rw [ eq_sub_iff_add_eq, add_comm,
+                        ← eq_sum_orthogonal_projection_self_orthogonal_complement U x,
+                        continuous_linear_map.id_apply ],
+     simp only [this],
+     clear this,
+     simp only [ continuous_linear_map.id_apply, map_sub, sub_eq_self,
+                 add_subgroup_class.coe_sub, sub_eq_zero,
+                 ← subspace_is_invariant_iff_ortho_proj_mul_T_mul_ortho_proj_eq_T_mul_ortho_proj],
+     exact ⟨λ ⟨h1,h2⟩ x, by
+            simp only [h2 x];
+            exact orthogonal_projection_eq_self_iff.mpr
+              (h1 (orthogonal_projection_fn_mem x)),
+            λ h, ⟨λ u h', by specialize h u;
+                            simp only [orthogonal_projection_eq_self_iff.mpr h'] at h;
+                            rw submodule.mem_comap;
+                            exact orthogonal_projection_eq_self_iff.mp h,
+                  λ x, by simp only [← h, orthogonal_projection_mem_subspace_eq_self]⟩⟩,
+   end
+
+-- Given an invertible operator, multiplying it by its inverse gives the identity.
+lemma mul_inv_operator (T : V →L[ℂ] V) [invertible T] :
+ T.inverse * T = 1 ∧ T * T.inverse = 1 :=
```

Comment:
This should be two separate lemmas; and should probably go in the same file as where `.inverse` is defined.

Reply:
I separated them, but I wasn't able to put them in the same file as where `.inverse` is defined - my computer wasn't able to run the change and kept crashing. Not sure if what I'm doing is the problem or not?